packages/components/src/Simple.tsx

```diff
@@ -3,92 +3,124 @@ import {
-  substanceString: string;
-  styleString: string;
+  domain: string;
+  substance: string;
+  style: string;
   variation: string;
-  initState?: PenroseState;
+  interactive?: boolean; // considered true by default
+  animate?: boolean; // considered false by default
 }
 
-interface ISimpleState {
-  state?: PenroseState;
-}
+class Simple extends React.Component<ISimpleProps> {
+  readonly canvasRef = React.createRef<HTMLDivElement>();
+  penroseState: PenroseState | undefined = undefined;
+  timerID: number | undefined = undefined; // for animation
 
-class Simple extends React.Component<ISimpleProps, ISimpleState> {
-  public readonly canvasRef = React.createRef<HTMLDivElement>();
-  constructor(props: ISimpleProps) {
-    super(props);
-    this.state = { state: undefined };
-  }
-  getInitState = async (
-    dsl: string,
-    sub: string,
-    sty: string,
-    variation: string
-  ): Promise<PenroseState | undefined> => {
-    const compilerResult = compileTrio({
-      substance: sub,
-      style: sty,
-      domain: dsl,
-      variation,
-    });
+  compile = async (): Promise<void> => {
+    this.penroseState = undefined;
+    const compilerResult = compileTrio(this.props);
     if (compilerResult.isOk()) {
       // resample because initial sampling did not use the special sampling seed
-      const initState: PenroseState = resample(
-        await prepareState(compilerResult.value)
-      );
-      const stepped = stepUntilConvergence(initState);
+      this.penroseState = resample(await prepareState(compilerResult.value));
+    } else {
+      console.log(showError(compilerResult.error));
+    }
+  };
+
+  converge = async (): Promise<void> => {
+    if (this.penroseState) {
+      const stepped = stepUntilConvergence(this.penroseState);
       if (stepped.isOk()) {
-        return stepped.value;
+        this.penroseState = stepped.value;
       } else {
         console.log(showError(stepped.error));
       }
-    } else {
-      console.log(showError(compilerResult.error));
     }
   };
+
+  tick = () => {
+    if (
+      this.props.animate &&
+      this.penroseState &&
+      !stateConverged(this.penroseState)
+    ) {
+      this.penroseState = stepState(this.penroseState, 1);
+      this.renderCanvas();
+    }
+  };
+
   componentDidMount = async () => {
-    const {
-      substanceString,
-      styleString,
-      domainString,
-      variation,
-    } = this.props;
-    const state: PenroseState | undefined = await this.getInitState(
-      domainString,
-      substanceString,
-      styleString,
-      variation
-    );
-    this.setState({ state });
-    this.renderCanvas(state);
+    await this.compile();
+    if (!this.props.animate) {
+      await this.converge();
+    }
+    this.renderCanvas();
+    this.timerID = window.setInterval(() => this.tick(), 1000 / 60);
   };
 
-  renderCanvas = async (state: PenroseState | undefined) => {
+  componentDidUpdate = async (prevProps: ISimpleProps) => {
+    if (
```

Comment:
A little surprised that prop updates need to be handled this manually. Wondering if we are leveraging the React lifecycle fully.